### PR TITLE
add highlight to starting bracket

### PR DESCRIPTION
**Pull request by pwjablonski**
@outoftime Thought I'd play around with this. I updated the highlight color to make it more visible and made it so the initial tag with the cursor is highlighted in addition to the matching tag. This works for any html element, } and ).

I used the ace editor functionality using the token and bypassed the redux store storing the maker id in the editor session. Not sure if this is the best implementation. 

I thought this type of feature aligns with the mission of popcode but love to hear your thoughts 

screen shot 2018-02-24 at 11 09 34 pm
screen shot 2018-02-24 at 11 11 46 pm
screen shot 2018-02-24 at 11 11 32 pm

### CODE REVIEW

File: src/components/Editor.jsx [hunk 1]
```diff
@@ -117,6 +126,19 @@ class Editor extends React.Component {
     session.on('change', () => {
       this.props.onInput(this._editor.getValue());
     });
+    session.selection.on('changeCursor', () => {
+      const {row, column} = this._editor.getCursorPosition();
+      const token = this._editor.session.getTokenAt(row, column);
+      this._editor.session.removeMarker(this._editor.session.$highlightMarker);
```
[the remark this answers is not in the document]
Comment (pwjablonski) [reply src/components/Editor.jsx]: Have seen any great documentation. This is what I've been going off ...

https://ace.c9.io/api/edit_session.html#EditSession.addMarker

addMarker as defined in ace repository 
https://github.com/ajaxorg/ace/blob/5d7ce2e54fe108c9b49190467a2cf24aee26e677/lib/ace/edit_session.js#L599

highlightBracket as defined in ace repository  
https://github.com/ajaxorg/ace/blob/master/lib/ace/editor.js#L520

File: src/components/Editor.jsx [hunk 2]
```diff
@@ -117,6 +126,19 @@ class Editor extends React.Component {
...
+      if (token && includes(BRACKETTYPES, token.type)) {
+        this._editor.session.$highlightMarker =
+          this._editor.session.addMarker(
+            new Range(row, token.start, row, token.start + token.value.length),
+            'ace_bracket',
```
[the remark this answers is not in the document]
Comment (pwjablonski) [reply src/components/Editor.jsx]: ace_bracket is built in. We are just adding additional styling.

File: src/components/Editor.jsx [hunk 3]
```diff
@@ -117,6 +126,19 @@ class Editor extends React.Component {
     session.on('change', () => {
       this.props.onInput(this._editor.getValue());
     });
+    session.selection.on('changeCursor', () => {
+      const {row, column} = this._editor.getCursorPosition();
```
[the remark this answers is not in the document]
Comment (pwjablonski) [reply src/components/Editor.jsx]: Indeed

**Comment by pwjablonski** [comment]
@outoftime cool!
 
- not tied to the green at all. Just wanted something that stands out more than the existing gray. thoughts @kaylee42?
- Agreed - true this is not trivial as we'd likely have to override / rework the existing highlight matching brackets function.

File: src/components/Editor.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (pwjablonski) [reply src/components/Editor.jsx]: agreed

**Comment by kaylee42** [comment]
ooooooo i love this idea! i wasn't even thinking about the highlight bracket when i proposed updating the color scheme, but i agree it needs to stand out more. i haven't had a chance to come up with a new color scheme (hoping to do that Sunday - so far the ones I've found that are color-blind friendly assume a dark background, so gonna have to make some adjustments) so... i can commit to having a palette proposal by the weekend that we can pick a highlight color out of if that works

File: src/components/Editor.jsx [hunk 4]
```diff
@@ -121,23 +121,27 @@ class Editor extends React.Component {
     }
   }
 
+  _addMarkerToOpeningBracket() {
+    const {row, column} = this._editor.getCursorPosition();
+    const token = this._editor.session.getTokenAt(row, column);
+    this._editor.session.removeMarker(this._editor.session.$highlightMarker);
```
Comment (outoftime) [inline src/components/Editor.jsx]: Much nicer refactored into a method?

One thing I’m a little confused about—my read of the code here is that we are removing one marker, and then adding (at most) one marker on line 130. However, the effect of this code from playing around with it is that there are _two_ markers, for the beginning and end, which is not the case in the production code. Where’s the extra marker coming from?

Also, I noticed that if the code is in an invalid state and there is, for instance, an opening bracket with no matching closing bracket, the opening bracket is still highlighted. That doesn’t seem ideal. Is that reasonably easy to fix?

**Comment by kaylee42** [comment]
Alright so I ended up adjusting a pre-existing theme to pretty closely match the [12-color palette for color blindness](http://mkweb.bcgsc.ca/colorblind/img/colorblindness.palettes.simple.png) (rather than the 7 cause I can't use the really light colors since they're contrast failures). However, I'm not so worried about color contrast failures on the highlight brackets since the don't need to be legible, not visible. There are two colors left I think could fit this case well - the pink (#FA78FA), and orange (#FA7850). (doesn't have to be that exact hex color but something close to it) We might be able to get away with light teal (#14D2DC), though there's already a teal in there, but the teal isn't as in your face as the other two.

**Comment by pwjablonski** [comment]
Closing this PR, for now, can pick it back up when we have a clearer sense of the updates to the Editor Themes from #1359 and #1361

**Pull request closed without merging**